### PR TITLE
lib.getExe: Do not make assumptions about the main program

### PR DESCRIPTION
**Pull request by roberth**
Before this commit, getExe assumes that if `meta.mainProgram` is unset, it has a main program that's named after the package name.

While this is probable, it leads to a bad error when the assumption does not hold. If the user called `getExe` themselves, they might narrow down the location of the assumption quite easily, but if that's not the case, they'll have to go down the rabbit hole to figure out what went wrong.

For example, a NixOS module may use `lib.getExe` on a package-typed option which is then used in the system configuration. This then typically leads to a failure *after* deployment, which is bad, and it's quite likely that the user will debug the package output contents before digging through the NixOS module, which is bad.
Furthermore the `getExe` call is rather inconspicuous as it does not contain something like "/bin/foo", which is bad.
Also modules can be hard to read for a newbie, which is bad.

All of this can be avoided by requiring `meta.mainProgram`. Many packages already have the attribute, and I would expect 80% of `getExe` usages to be covered by 20% of packages, because plenty of packages aren't used with `getExe` anyway.

Finally we could make an effort to set `mainProgram` semi-automatically using `nix-index`.

## Description of changes

Adds a warning to the unreliable behavior.

```console
[nixpkgs]$ nix repl .
nix-repl> lib.getExe (hello // { meta = {}; })
trace: warning: getExe: Package hello does not have the meta.mainProgram attribute. We'll assume that the main program has the same name for now, but this behavior is deprecated, because it leads to surprising errors when the assumption does not hold. If the package has a main program, please set `meta.mainProgram` in its definition to make this warning go away. Otherwise, if the package does not have a main program, or if you don't control its definition, specify the full path to the program, such as "${lib.getBin foo}/bin/bar".
"/nix/store/nj2460zk9fjdl9yyglsi2nh4lp78na24-hello-2.12.1/bin/hello"

```

## Context

- Introduced in https://github.com/NixOS/nixpkgs/pull/167247
  - [Short conversation leading to this PR](https://github.com/NixOS/nixpkgs/pull/167247/files#r1279561455)

- Tangentially related to the point of off-topic: https://github.com/NixOS/nixpkgs/pull/173675

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by drupol — APPROVED** [review]
LGTM and I support this.

**Review by Lassulus — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by AndersonTorres** [comment]
Are you telling me `mainProgram` is null by default??

What about this? https://ryantm.github.io/nixpkgs/stdenv/meta/#var-meta-mainProgram

**Comment by roberth** [comment]
> Are you telling me `mainProgram` is null by default??

No, it's unset.

> What about this? https://ryantm.github.io/nixpkgs/stdenv/meta/#var-meta-mainProgram

That's clearly wrong.

File: nixos/modules/programs/oddjobd.nix [hunk 1]
```diff
@@ -10,6 +10,11 @@ in
   };
 
   config = lib.mkIf cfg.enable {
+    assertions = [
```
Comment (RaitoBezarius) [inline nixos/modules/programs/oddjobd.nix:13]: I really like this pattern!

**Comment by figsoda** [comment]
One concern I have with this is that we will no longer have something equivalent to `nix run`.

As an alternative, we can do something like this:
- Have a way to explicitly disable `mainProgram` for packages that do not have a main program, something like `mainProgram = null` which should throw a meaningful error message
- Make ofborg check that `"${lib.getBin pkg}/bin/${pkg.meta.mainProgram or (lib.getName pkg)}"` is executable
	- the check should be skipped when `mainProgram` is explicitly disabled 
	- maybe this can be implemented as a default `passthru.tests` in `stdenv.mkDerivation`
- The same check can be added to NixOS options that accept a package, so things fail at build time instead of runtime

This makes sure that (at least for new packages) `lib.getExe` always works, while not requiring any treewide setting of `meta.mainProgram`

**Comment by roberth** [comment]
> One concern I have with this is that we will no longer have something equivalent to `nix run`.

It can't be exactly equivalent to `nix run` because that commands has some extra behavior such as prepending to `PATH`.



> * Have a way to explicitly disable `mainProgram` for packages that do not have a main program, something like `mainProgram = null` which should throw a meaningful error message when used with `lib.getExe` or `nix run`

This could be used for representing "this is not a data incompleteness problem", but I don't know if it makes the user experience that much better. Maybe?

I would recommend against using this format to reinstate the `name`/`pname` default behavior, because doing so does not lead to an improvement in nixpkgs data quality.


> * maybe this can be implemented as a default `passthru.tests` in `stdenv.mkDerivation`

This has some characteristics similar to the pkg-config metadata. [This post](https://discourse.nixos.org/t/adding-pkg-config-metadata-to-nixpkgs-packages/25281) aggregates some links that show how that was discussed and implemented.

I'm not a huge fan of making `mkDerivation` even more complicated, and perhaps this could be a separate "package set" containing the checks.

For sure there's more than one way to go about this.

Pet peeve: it's not `passthru` if it's an automatically added test. I wish we could just call it the `tests` _package attribute_, where `passthru` is just an implementation detail of how you create it with `mkDerivation`.

> * The same check can be added to NixOS options that accept a package, so things fail at build time instead of runtime

Not sure how this would work. `types.package` options have a value that's just the store path without `/bin/anything`.



> treewide setting of `meta.mainProgram`

This is valuable data even if it's a fractional bit of information for many packages.

**Comment by AndersonTorres** [comment]
I am not a fan of such "assumptions". Being explicit is more verbose but better at the long run.

File: pkgs/development/interpreters/python/cpython/default.nix [hunk 2]
```diff
@@ -572,5 +572,6 @@ in with passthru; stdenv.mkDerivation {
     license = licenses.psfl;
     platforms = platforms.linux ++ platforms.darwin;
     maintainers = with maintainers; [ fridh ];
+    mainProgram = "python3";
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/python/cpython/default.nix:575]: This is not quite correct, fix in #247955

Comment (roberth) [reply pkgs/development/interpreters/python/cpython/default.nix:575]: It preserved the prior behavior of `lib.getExe` on python, by virtue of being equal to the `pname` value (except for `python3Minimal`, where `"python3"` is a fix).

I have no opinion on what the `mainProgram` for the python interpreters should be.

**Pull request merged**